pkg/workqueue/dispatcher/dispatcher.go

```diff
@@ -114,6 +114,12 @@ func HandleAsync(ctx context.Context, wq workqueue.Interface, concurrency int, f
 					if err := oip.Deadletter(ctx); err != nil {
 						return fmt.Errorf("fail(after reaching max retries) = %w", err)
 					}
+				} else if d := workqueue.GetNonRetriableDetails(err); d != nil {
+					clog.InfoContextf(ctx, "Key %q is marked as non-retriable - reason: %s, err: %v", oip.Name(), d.GetMessage(), err)
+					// If the error is marked as non-retriable, we should not requeue it.
+					if err := oip.Complete(ctx); err != nil {
+						return fmt.Errorf("complete(after non-retriable error) = %w", err)
+					}
```

Comment:
We need to be really careful about how we use this as we could shed some of the key resilience the workqueue gives us.